bfloat16: vectorized clamp, clamp_min and clmap_max

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35172 bfloat16: enable basic math function
* #35092 bfloat16: vectorized unary ops
* **#35082 bfloat16: vectorized clamp, clamp_min and clmap_max**

Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

XiaobingSuper committed:
bfloat16: vectorized clamp, clamp_min and clmap_max

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper commented:
@uyongw, @Jianhui-Li, @jgong5, @hongzhen1

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"




[ghstack-poisoned]

ngimel commented on test/test_torch.py:
```diff
@@ -6479,6 +6479,30 @@ def test_clamp(self, device):
 
+    @onlyCPU
+    def test_clamp_bfloat16(self, device):
+        a = torch.randn(4, 4, dtype=torch.float, device=device)
```
This test won't trigger vectorized pass, because there are 16 elements in Vec256<Bfloat16> and 2 iterations are required for a vectorized loop.

ngimel commented:
Here, and in subsequent PRs, it is preferred to test using the new testing framework (tests generated using `instantiate_device_type_tests`, look at the bottom of test_torch.py. I suggest adding your tests to `TestTensorDeviceOps`, here https://github.com/pytorch/pytorch/compare/master...ngimel:prec_cpu_test?expand=1 I modified it so that _cpu_types can be added to the tests. In most tests it will remain empty, and thus they will be skipped on the cpu, but for operations that you want to test you should set _cpu_types to torch.bfloat16, then a test will be generated that compares result of bfloat16 operation with float32. TestTensorDeviceOps already has entries for all the ops you are testing in this and 2 following PRs, if it does not, feel free to add what you need to tensor_op_tests.

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"


Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"


Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"


Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

[ghstack-poisoned]

ngimel commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -505,7 +505,7 @@ class dtypes(object):
     # Python 3 allows (self, *args, device_type='all').
     def __init__(self, *args, **kwargs):
-        assert args is not None and len(args) != 0, "No dtypes given"
+        # assert args is not None and len(args) != 0, "No dtypes given"
```
Nit: remove this commented out line

XiaobingSuper replied:
done.

ngimel reviewed (approved):
Approving, please remove commented out line.

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"


Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

[ghstack-poisoned]

XiaobingSuper committed:
Update on "bfloat16: vectorized clamp, clamp_min and clmap_max"


Differential Revision: [D20721148](https://our.internmc.facebook.com/intern/diff/D20721148)

[ghstack-poisoned]
